src/python/pants/backend/awslambda/common/awslambda_common_rules.py

```diff
@@ -79,4 +79,4 @@ async def create_awslambda(
 
 
 def rules():
-    return [create_awslambda]
+    return register_rules()
```

Comment:
Thoughts on inverting this, and calling it from outside the module? Ie, having an importer call with the module name:
```
register_rules(pants.backend.awslambda.common.awslambda_common_rules)
```

Would have two advantages I think:
1. no uncertainty about what happens when called with no args (although that's not _too_ magical)
2. no need for a convention inside of each module to export rules: things are mostly consistent, but avoiding the need for the convention entirely would be great.

It looks like to make it work though, `UnionRule` could maybe move to being a class decorator... ie:
```
@union
class MyUnion: pass

@union_member(MyUnion)
class MyUnionMember: pass
```

...so maybe better as a followup step.

Reply:
> Thoughts on inverting this, and calling it from outside the module?

The API supports it and that was my intention, but its not easy right now so took this smaller step.
Since we hack RootRules in to support tests in our production rules() lists (not cool), using this function from outside the module does not catch those and foces you to register those.

> It looks like to make it work though, UnionRule could maybe move to being a class decorator... ie:

I have an old branch that instead just forces union members to be subclasses of the @union decorated (potentially just marker) type. That will kill all member boilerplate - you just subclass which should feel like it makes sense to rule authors opting into the `test` goal say from a pure Python perspective, Pants engine aside.

Reply:
I like the idea of `@union_member`, although I don't think that would work with plugin fields: https://github.com/pantsbuild/pants/pull/10469

I also don't think that we want plugin fields to have to subclass the target they're being used for. A key part of the Target API is that a single field definition may be used across multiple targets. Also, fields already subclass a field template. So, we would end up with:

```python
class Example(IntField, PythonLibrary.PluginField, PythonTests.PluginField):
   alias = "example"
```

Reply:
PluginField should not be using Union IMO. If we instead `Target.register_plugin_field(Field) -> Rule` then Rule can be whatever type we want and e don't need to bastardize UnionRule. Hopefully everyone realizes that by bastardizing Unions in this way, if you look at the action graph you get spammed with a bunch of fake entries for non-rule edges.

Reply:
Union has a very specific purpose: allow a Get to retrieve a given product type from an unknown provider - ie a rule set forthcoming that can provide that product type given an unknown input type at the time of writing the top-level goal. Thats pretty darn specific. Not everything is a union, just that.